mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -191,6 +191,8 @@ public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
 			for ( int i = 0; i < searchObjects.length; i++ ) {
 				if ( dbObject.get( ID_FIELDNAME ).equals( searchObjects[i] ) ) {
 					tuples.set( i, createTuple( keys[i], tupleContext, dbObject ) );
```

Comment:
Thinking about it, should we better consume the cursor into a map and then drive result set creating by iterating the requested keys and looking up the corresponding result.

In the current approach we'll have n * n/2 loop iterations on average (it used to be n * n, so it's better already), an n can go up to 1000 as per @emmanuelbernard. Using a temporary map, it'd be n + n (once iterating the cursor, once the incoming keys), at the cost of the map allocation. Of course we should not optimize prematurely, but the n * n growth is scary. @Sanne, any thoughts?